File: mac-app/QuartzTeachers/Models/FolderPathRewriter.swift
import Foundation

/// Pointing every link that names a FOLDER at that folder's new name.
///
/// The sibling of `WikiLinkRewriter`, and deliberately a separate type: that
/// one rewrites the name of a PAGE, which is what Obsidian's usual
/// `[[Unit 2, Day 3]]` carries. This one rewrites a path SEGMENT, which is
/// what appears when a link is qualified — `[[Tasks/Quiz 1]]`, a full vault
/// path, or a Markdown link written by Obsidian's other link style.
///
/// **Most links need nothing done to them, and that is the important fact.**
/// Obsidian resolves `[[Quiz 1]]` by searching the vault, so moving the folder
/// it lives in leaves the link working. The `TODO.md` entry that deferred this
/// feature worried that a folder rename would strand every link pointing into
/// it; it does not, and the cases that DO break are the qualified ones handled
/// here. That is why renaming a folder is a far smaller risk than renaming a
/// page, and why this could be built without the undo a page rename would
/// need.
///
/// ## What is handled
///
/// * `[[Tasks/Quiz 1]]` and `![[Tasks/diagram.png]]`
/// * `[[ICS3U/section1/Tasks/Quiz 1]]` — a full vault path, so ANY segment is
///   matched, not only the first
/// * `[[Tasks/Quiz 1#Marking|the quiz]]` — the alias and the heading are the
///   teacher's own words and are never touched
/// * `[the quiz](Tasks/Quiz%201.md)` and `![](Tasks/diagram.png)` — Markdown
///   style, with or without percent-encoded spaces, and with a leading `./`
///
/// A Markdown destination ends at the first space, so a new name that has one
/// is percent-encoded on the way in and a wikilink's is not. That rule, and
/// what it is measured against, is in `spelled(_:likeThe:in:)`.
///
/// ## What is NOT handled, on purpose
///
/// A segment is replaced only when it matches the whole folder name. A folder
/// called `Tasks` does not rewrite `Extra Tasks/`, and a page whose own NAME is
/// `Tasks.md` is left alone — the match stops at the last `/`, so a file name
/// is never a candidate.
///
/// **Links to the web and absolute paths are refused explicitly**, by
/// `pointsOutsideTheCourse`, and NOT because "nothing in them is a segment of
/// this course's tree" — which is what this comment used to claim and is
/// exactly wrong. `https://example.com/Tasks/handout.pdf` has `Tasks` sitting
/// in it as an ordinary segment, and the walk below is blind to what a path
/// means, so a rename used to repoint that link at a page on somebody else's
/// website. Found by adversarial review, 2026-09-01.
enum FolderPathRewriter {

    // MARK: - Nested types

    /// Which of the two link styles a target was written in.
    ///
    /// The rewriter has to know, because the two spell the SAME folder name
    /// differently. `[[All Tasks/Quiz 1]]` is exactly how Obsidian writes a
    /// wikilink whose folder has a space in it, while `](All Tasks/Quiz 1.md)`
    /// is not a link at all — a Markdown destination ends at the first space.
    /// See `spelled(_:likeThe:in:)`, where the difference is decided.
    nonisolated enum LinkStyle {

        case wikiLink
        case markdown

        // MARK: - Computed properties

        /// The pattern that finds this style's targets.
        var pattern: String {
            switch self {
            case .wikiLink:
                return FolderPathRewriter.wikiLinkPattern
            case .markdown:
                return FolderPathRewriter.markdownLinkPattern
            }
        }

        /// The compiled form of that pattern, built once for the whole run.
        var expression: NSRegularExpression? {
            switch self {
            case .wikiLink:
                return FolderPathRewriter.wikiLinkExpression
            case .markdown:
                return FolderPathRewriter.markdownLinkExpression
            }
        }
    }

    // MARK: - Stored properties

    /// An optional `!`, the opening brackets, then the target — which runs up
    /// to the first `]`, `|` or `#`, so an alias, a heading and a block
    /// reference stay where they are. `WikiLinkRewriter`'s pattern, and the
    /// same one on purpose: two link finders that disagreed about what a link
    /// is would rewrite different halves of the same vault.
    nonisolated static let wikiLinkPattern: String = #"(!?\[\[)([^\]|#]+)"#

    /// A Markdown link or embed's target: everything between `](` and the
    /// closing bracket. Titles (`](path "title")`) are left in place because
    /// the path is taken only up to the first space.
    nonisolated static let markdownLinkPattern: String = #"(\]\()([^)\s]+)"#

    /// Compiled ONCE, not per file. A rename walks every page in the course
    /// and each page used to build four of these; the cost is small but it is
    /// the kind of per-file work that makes an O(files) walk worse than it
    /// needs to be, and these patterns are constants.
    nonisolated private static let wikiLinkExpression: NSRegularExpression? =
        try? NSRegularExpression(pattern: wikiLinkPattern)

    nonisolated private static let markdownLinkExpression: NSRegularExpression? =
        try? NSRegularExpression(pattern: markdownLinkPattern)

    /// The name percent-encoded so that Quartz gets the folder's real name
    /// back out of it.
    ///
    /// **The allowed set is not `urlPathAllowed`, and not .NET's
    /// `Uri.EscapeDataString` either — both were tried and both are wrong
    /// here.** Quartz v4.5.0 resolves an internal link with `decodeURI`
    /// (`quartz/util/path.ts`, `transformInternalLink`), and `decodeURI`
    /// deliberately leaves the reserved set `; / ? : @ & = + $ , #` still
    /// encoded. It then slugs what comes back, turning `&` into `-and-` and
    /// `%` into `-percent` (`sluggify`, same file). So a folder called
    /// `Tasks & Quizzes` written as `Tasks%20%26%20Quizzes` decodes to
    /// `Tasks %26 Quizzes` and slugs to `Tasks--percent26-Quizzes`, while the
    /// real folder slugs to `Tasks--and--Quizzes`: a 404 on the published
    /// site, and invisible in Obsidian, which decodes `%26` perfectly well.
    /// Measured against the running image on 2026-09-06.
    ///
    /// The set below is therefore what JavaScript's `encodeURI` leaves alone,
    /// minus the four that a Markdown destination or a slug cannot hold:
    /// `(` and `)` close the destination, `#` starts a heading and `?` a
    /// query. `/` and `:` are left out too — the rename sheet refuses both, so
    /// they cannot arrive, and encoding is the safer of the two ways to be
    /// wrong if that ever changes. Everything else — the space, `%`, the
    /// quotes and brackets, and every non-ASCII letter — is encoded, and
    /// `decodeURI` gives all of it back.
    nonisolated private static let charactersThatSurviveQuartzUndecoded: CharacterSet = CharacterSet(
        charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789;,@&=+$-_.!~*'"
    )

    // MARK: - Functions

    /// The text with every qualified link to `oldName` pointing at `newName`.
    nonisolated static func rewriting(_ text: String, folderNamed oldName: String, to newName: String) -> String {
        let trimmedOld: String = oldName.trimmingCharacters(in: .whitespaces)
        let trimmedNew: String = newName.trimmingCharacters(in: .whitespaces)
        if trimmedOld.isEmpty || trimmedNew.isEmpty || trimmedOld == trimmedNew {
            return text
        }
        let afterWikiLinks: String = rewritingTargets(
            in: text, written: .wikiLink, folderNamed: trimmedOld, to: trimmedNew
        )
        return rewritingTargets(
            in: afterWikiLinks, written: .markdown, folderNamed: trimmedOld, to: trimmedNew
        )
    }

    /// How many qualified links in this text name the folder. Used to report
    /// what a rename touched, and to skip writing a file nothing changed in.
    nonisolated static func countReferences(to folderName: String, in text: String) -> Int {
        let trimmed: String = folderName.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty {
            return 0
        }
        var total: Int = 0
        for style in [LinkStyle.wikiLink, LinkStyle.markdown] {
            for target in targets(in: text, written: style) {
                if pathNames(trimmed, in: target) {
                    total += 1
                }
            }
        }
        return total
    }

    // MARK: - Private helpers

    /// Every link target in the text, for one of the two link styles.
    nonisolated private static func targets(in text: String, written style: LinkStyle) -> [String] {
        guard let expression = style.expression else {
            return []
        }
        let whole: NSRange = NSRange(text.startIndex..<text.endIndex, in: text)
        var found: [String] = []
        for match in expression.matches(in: text, range: whole) {
            if let targetRange = Range(match.range(at: 2), in: text) {
                found.append(String(text[targetRange]))
            }
        }
        return found
    }

    /// The text with one link style's targets rewritten.
    ///
    /// Walks the matches in order and copies the text between them, rather
    /// than replacing in place: a replacement changes the string's length, and
    /// ranges found beforehand would then point at the wrong characters.
    nonisolated private static func rewritingTargets(
        in text: String, written style: LinkStyle, folderNamed oldName: String, to newName: String
    ) -> String {
        guard let expression = style.expression else {
            return text
        }
        let whole: NSRange = NSRange(text.startIndex..<text.endIndex, in: text)
        let matches: [NSTextCheckingResult] = expression.matches(in: text, range: whole)
        if matches.isEmpty {
            return text
        }

        var result: String = ""
        var carriedTo: String.Index = text.startIndex
        for match in matches {
            guard let matchRange = Range(match.range, in: text),
                  let openingRange = Range(match.range(at: 1), in: text),
                  let targetRange = Range(match.range(at: 2), in: text) else {
                continue
            }
            let target: String = String(text[targetRange])
            let rewritten: String = retargeting(target, folderNamed: oldName, to: newName, written: style)
            if rewritten == target {
                continue
            }
            result.append(contentsOf: text[carriedTo..<matchRange.lowerBound])
            result.append(String(text[openingRange]))
            result.append(rewritten)
            carriedTo = matchRange.upperBound
        }
        result.append(contentsOf: text[carriedTo...])
        return result
    }

    /// One link target with any segment naming the folder replaced.
    ///
    /// The LAST segment is the page or file and is never a candidate, so a
    /// page called `Tasks.md` survives a rename of the `Tasks` folder.
    nonisolated private static func retargeting(
        _ target: String, folderNamed oldName: String, to newName: String, written style: LinkStyle
    ) -> String {
        if !target.contains("/") || pointsOutsideTheCourse(target) {
            return target
        }
        var segments: [String] = []
        for segment in target.split(separator: "/", omittingEmptySubsequences: false) {
            segments.append(String(segment))
        }
        var changed: Bool = false
        var rebuilt: [String] = []
        for index in 0..<segments.count {
            let segment: String = segments[index]
            let isLastSegment: Bool = (index == segments.count - 1)
            if !isLastSegment && matches(segment, name: oldName) {
                rebuilt.append(spelled(newName, likeThe: segment, in: style))
                changed = true
            } else {
                rebuilt.append(segment)
            }
        }
        if !changed {
            return target
        }
        return rebuilt.joined(separator: "/")
    }

    /// Whether a link target leads somewhere this rename has no business
    /// touching: the web, or an absolute path on the teacher's machine.
    ///
    /// **Found by review rather than by use, and it was a real bug.** The
    /// segment walk below is blind to what a path MEANS, so
    /// `https://example.com/Tasks/handout.pdf` had `Tasks` sitting in it as an
    /// ordinary segment, and renaming a folder called `Tasks` silently
    /// repointed the link at a page on somebody else's website. Folders called
    /// `Resources`, `Files`, `Notes` or `Assignments` make that likely rather
    /// than exotic. Nothing in a course's own tree is reached by an absolute
    /// path or a URL, so refusing both costs nothing.
    nonisolated private static func pointsOutsideTheCourse(_ target: String) -> Bool {
        if target.hasPrefix("/") || target.hasPrefix("#") {
            return true
        }
        // A scheme — http:, https:, mailto:, obsidian:, file: — is anything
        // before a colon that comes ahead of the first slash. Tested that way
        // rather than against a list of schemes, because the list is open and
        // a missed one silently rewrites somebody's link.
        guard let colon = target.firstIndex(of: ":") else {
            return false
        }
        if let slash = target.firstIndex(of: "/") {
            return colon < slash
        }
        return true
    }

    /// Whether a target names this folder in any segment but its last.
    nonisolated private static func pathNames(_ folderName: String, in target: String) -> Bool {
        if !target.contains("/") || pointsOutsideTheCourse(target) {
            return false
        }
        var segments: [String] = []
        for segment in target.split(separator: "/", omittingEmptySubsequences: false) {
            segments.append(String(segment))
        }
        for index in 0..<(segments.count - 1) {
            if matches(segments[index], name: folderName) {
                return true
            }
        }
        return false
    }

    /// Whether one path segment IS this folder, allowing for the percent
    /// encoding Obsidian writes into Markdown-style links, and matching case
    /// insensitively the way Obsidian resolves names.
    nonisolated private static func matches(_ segment: String, name: String) -> Bool {
        let plain: String = segment.trimmingCharacters(in: .whitespaces)
        if plain.caseInsensitiveCompare(name) == .orderedSame {
            return true
        }
        if let decoded = plain.removingPercentEncoding {
            return decoded.caseInsensitiveCompare(name) == .orderedSame
        }
        return false
    }

    /// How the new name is spelled inside this particular link.
    ///
    /// **"Spell it the way the old segment was spelled" is the obvious rule
    /// and it is wrong**, which is what this used to do. A Markdown link's
    /// destination ends at the first SPACE — the pattern above is literally
    /// `[^)\s]+` — so renaming `Tasks` to `All Tasks` turned
    /// `[q](Tasks/Quiz%201.md)` into `[q](All Tasks/Quiz%201.md)`, which
    /// neither Obsidian nor Quartz can follow. The old segment `Tasks` carries
    /// no `%`, so nothing here escaped anything; the `%20` in that example
    /// belongs to the FILE name, which is what made it easy to miss by eye.
    /// Every Markdown-style link into the folder broke, in the teacher's own
    /// pages, and nothing said so. Found on Windows by adversarial review,
    /// 2026-09-06, and reported to this side as a shared defect rather than a
    /// port error — see `contracts/shared-rules.json` →
    /// `specialNames.renameFolder.linkRewriting`.
    ///
    /// So: **in a Markdown link, escape when the NEW name needs it, whatever
    /// the old segment looked like. In a wikilink, keep the plain spelling** —
    /// `[[All Tasks/Quiz 1]]` is how Obsidian writes a wikilink containing a
    /// space, and escaping there would be the mirror-image mistake.
    ///
    /// The old rule is kept as a second reason to escape rather than replaced
    /// by the new one: a segment that ARRIVED percent-encoded goes back
    /// percent-encoded, so a link a teacher already had keeps the shape it had.
    nonisolated private static func spelled(_ name: String, likeThe segment: String, in style: LinkStyle) -> String {
        if style == .markdown && wouldBreakAMarkdownTarget(name) {
            return percentEncoded(name)
        }
        if wasPercentEncoded(segment) {
            return percentEncoded(name)
        }
        return name
    }

    /// Whether this name, dropped into a Markdown destination as it stands,
    /// would end the destination early or fail to parse.
    ///
    /// Whitespace and round brackets are the two that close a destination.
    /// A lone `%` is the third and is less obvious: Quartz resolves an
    /// internal link by calling JavaScript's `decodeURI` on it, and
    /// `decodeURI("10%/Quiz.md")` throws rather than returning anything.
    nonisolated private static func wouldBreakAMarkdownTarget(_ name: String) -> Bool {
        for character in name {
            if character.isWhitespace || character == "(" || character == ")" || character == "%" {
                return true
            }
        }
        return false
    }

    /// The name percent-encoded so that Quartz gets the folder's real name
    /// back out of it. See `charactersThatSurviveQuartzUndecoded` above for
    /// why the allowed set is neither `urlPathAllowed` nor the whole reserved
    /// set — both were tried and both are wrong here.
    nonisolated private static func percentEncoded(_ name: String) -> String {
        if let encoded = name.addingPercentEncoding(
            withAllowedCharacters: charactersThatSurviveQuartzUndecoded
        ) {
            return encoded
        }
        return name
    }

    /// Whether this segment arrived percent-encoded.
    ///
    /// A segment with no `%` in it never is, and one whose `%` does not begin
    /// a valid escape is treated as plain text rather than as encoding —
    /// `removingPercentEncoding` answers `nil` there, and reading that as
    /// "encoded" would be the opposite of what it means.
    nonisolated private static func wasPercentEncoded(_ segment: String) -> Bool {
        if !segment.contains("%") {
            return false
        }
        guard let decoded = segment.removingPercentEncoding else {
            return false
        }
        return decoded != segment
    }
}

File: windows-app/Plantoir.Core/Models/FolderPathRewriter.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace Plantoir.Core.Models;

/// <summary>
/// Pointing every link that names a FOLDER at that folder's new name.
///
/// <para><b>Most links need nothing done to them, and that is the important
/// fact.</b> Obsidian resolves <c>[[Quiz 1]]</c> by searching the vault, so
/// moving the folder it lives in leaves the link working. Renaming a folder is
/// therefore a far smaller risk than renaming a page, and is why this could be
/// built without an undo: only QUALIFIED links break, and those are what this
/// handles.</para>
///
/// <para><b>What is handled</b> — <c>[[Tasks/Quiz 1]]</c> and
/// <c>![[Tasks/diagram.png]]</c>; a full vault path such as
/// <c>[[ICS3U/section1/Tasks/Quiz 1]]</c>, so ANY segment is matched and not
/// only the first; <c>[[Tasks/Quiz 1#Marking|the quiz]]</c>, whose alias and
/// heading are the teacher's own words and are never touched; and Markdown
/// style, <c>[the quiz](Tasks/Quiz%201.md)</c>, with or without percent-encoded
/// spaces and with a leading <c>./</c>.</para>
///
/// <para><b>What is NOT handled, on purpose.</b> A segment is replaced only
/// when it matches the WHOLE folder name: a folder called <c>Tasks</c> does not
/// rewrite <c>Extra Tasks/</c>, and a page whose own name is <c>Tasks.md</c> is
/// left alone, because the match stops before the last <c>/</c> and a file name
/// is never a candidate. A rewriter that matched substrings would rename
/// folders the teacher never touched.</para>
///
/// <para><b>Links that point outside the course are refused explicitly</b>, and
/// NOT because "nothing in them is a segment of this course's tree" — that
/// reasoning is exactly wrong. <c>https://example.com/Tasks/handout.pdf</c> has
/// <c>Tasks</c> sitting in it as an ordinary segment, and the walk below is
/// blind to what a path MEANS, so without the refusal a rename would repoint
/// that link at a page on somebody else's website. Found by adversarial review
/// on the mac, 2026-09-01, and inherited here as a rule rather than as
/// code.</para>
/// </summary>
public static class FolderPathRewriter
{
    /// <summary>
    /// An optional <c>!</c>, the opening brackets, then the target — which runs
    /// up to the first <c>]</c>, <c>|</c> or <c>#</c>, so an alias, a heading
    /// and a block reference stay where they are.
    /// </summary>
    private static readonly Regex WikiLink =
        new(@"(!?\[\[)([^\]|#]+)", RegexOptions.Compiled);

    /// <summary>
    /// A Markdown link or embed's target: everything between <c>](</c> and the
    /// closing bracket. Titles (<c>](path "title")</c>) are left in place
    /// because the path is taken only up to the first space.
    /// </summary>
    private static readonly Regex MarkdownLink =
        new(@"(\]\()([^)\s]+)", RegexOptions.Compiled);

    /// <summary>A URL scheme: http:, https:, mailto:, obsidian: and friends.</summary>
    private static readonly Regex Scheme =
        new(@"^[A-Za-z][A-Za-z0-9+.\-]*:", RegexOptions.Compiled);

    /// <summary>
    /// The characters that go into a link AS THEY STAND when escaping runs.
    /// Everything else is percent-encoded as UTF-8.
    ///
    /// <para><b>This set was measured against the built site, not reasoned
    /// about, and <c>Uri.EscapeDataString</c> is the wrong tool for it.</b>
    /// Quartz resolves an internal link with JavaScript's <c>decodeURI</c>
    /// (<c>quartz/util/path.ts</c>, <c>transformInternalLink</c>), and
    /// <c>decodeURI</c> deliberately leaves the reserved set
    /// <c>; / ? : @ &amp; = + $ , #</c> STILL ENCODED — that is what
    /// distinguishes it from <c>decodeURIComponent</c>. <c>sluggify</c>, in the
    /// same file, then maps <c>&amp;</c> to <c>-and-</c> and <c>%</c> to
    /// <c>-percent</c> when it builds the address. So a folder called
    /// "Tasks &amp; Quizzes" written as <c>Tasks%20%26%20Quizzes</c> decodes to
    /// <c>Tasks %26 Quizzes</c> and slugs to <c>Tasks--percent26-Quizzes</c>,
    /// while the real folder slugs to <c>Tasks--and--Quizzes</c>: a 404 for
    /// students, and invisible to the teacher, because Obsidian decodes
    /// <c>%26</c> perfectly well. <c>Uri.EscapeDataString</c> keeps only
    /// <c>A-Za-z0-9-._~</c>, so it produces exactly that — and over-encodes
    /// <c>,</c> the same way, which is the one that will actually happen here:
    /// "Unit 1, Day 2" is this project's own naming pattern.</para>
    ///
    /// <para>The set is what JavaScript's <c>encodeURI</c> leaves alone, minus
    /// three — <c>(</c> and <c>)</c> close a destination, <c>#</c> starts a
    /// heading — and minus <c>/</c> and <c>:</c>, which the rename sheet
    /// refuses anyway. <c>?</c> IS in it: <c>sluggify</c> STRIPS a question
    /// mark from the real folder's name, so "Why Not?" slugs to
    /// <c>Why-Not</c>, an unescaped <c>Why%20Not?</c> slugs to the same and
    /// resolves, and the escaped <c>Why%20Not%3F</c> slugs to
    /// <c>Why-Not-percent3F</c> and 404s. A folder name may not contain a
    /// question mark on Windows, but this is a pure string transform and the
    /// rule is shared.</para>
    ///
    /// <para><b>Nothing is encoded unless escaping RUNS</b> — see
    /// <see cref="WouldBreakAMarkdownTarget"/>, or a segment that arrived
    /// percent-encoded. "Café" goes into a link as "Café"; "Café Notes" goes in
    /// as <c>Caf%C3%A9%20Notes</c>. Both resolve, and reading this set as
    /// "always encode everything else" is the way the two apps would write
    /// different text for the same rename.</para>
    ///
    /// <para>Written down once, in
    /// <c>contracts/shared-rules.json</c> →
    /// <c>specialNames.renameFolder.linkRewriting.escapingSet.leaveUnescaped</c>,
    /// and pinned character-for-character against this constant by
    /// <c>FolderPathRewriterTests</c>. Change it there, not here.</para>
    /// </summary>
    internal const string CharactersThatSurviveQuartzUndecoded =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789;,@&=+$-_.!~*'?";

    /// <summary>
    /// Every link in <paramref name="text"/> that names <paramref name="oldName"/>
    /// as a folder, pointed at <paramref name="newName"/>.
    /// </summary>
    public static string Rewritten(string text, string oldName, string newName)
    {
        if (string.IsNullOrEmpty(text)) return text;
        if (string.IsNullOrWhiteSpace(oldName) || string.IsNullOrWhiteSpace(newName)) return text;
        if (oldName.Equals(newName, StringComparison.Ordinal)) return text;

        string once = WikiLink.Replace(text, match => Replaced(match, oldName, newName, markdown: false));
        return MarkdownLink.Replace(once, match => Replaced(match, oldName, newName, markdown: true));
    }

    /// <summary>
    /// How many links point INTO this folder by name — what the "and N pages
    /// had links pointing into it" sentence counts. Only qualified links count;
    /// a bare <c>[[Quiz 1]]</c> does not point into anything.
    /// </summary>
    public static int Count(string text, string folderName)
    {
        if (string.IsNullOrEmpty(text) || string.IsNullOrWhiteSpace(folderName)) return 0;
        int found = 0;
        foreach (Regex pattern in new[] { WikiLink, MarkdownLink })
            foreach (Match match in pattern.Matches(text))
                if (NamesTheFolder(match.Groups[2].Value, folderName))
                    found++;
        return found;
    }

    private static string Replaced(Match match, string oldName, string newName, bool markdown)
    {
        string opening = match.Groups[1].Value;
        string target = match.Groups[2].Value;
        return opening + RewrittenTarget(target, oldName, newName, markdown);
    }

    /// <summary>
    /// One link target, with any SEGMENT naming the folder replaced.
    ///
    /// <para>The last segment is the page's own file name and is never a
    /// candidate, so a page called <c>Tasks.md</c> survives a rename of the
    /// folder <c>Tasks</c>.</para>
    /// </summary>
    private static string RewrittenTarget(string target, string oldName, string newName, bool markdown)
    {
        if (PointsOutsideTheCourse(target)) return target;

        var segments = new List<string>(target.Split('/'));
        if (segments.Count < 2) return target;   // nothing but a file name

        bool changed = false;
        for (int i = 0; i < segments.Count - 1; i++)
        {
            if (!SegmentIs(segments[i], oldName)) continue;
            segments[i] = Spelled(newName, wasEncoded: WasEncoded(segments[i]), markdown: markdown);
            changed = true;
        }
        return changed ? string.Join("/", segments) : target;
    }

    /// <summary>
    /// How the new name is spelled inside this particular link.
    ///
    /// <para>Keeping the style the link already used is the obvious rule and is
    /// WRONG for Markdown links on its own. A Markdown destination ends at the
    /// first space — the pattern above is literally <c>[^)\s]+</c> — so
    /// renaming <c>Tasks</c> to <c>All Tasks</c> turned
    /// <c>[q](Tasks/Quiz%201.md)</c> into <c>[q](All Tasks/Quiz%201.md)</c>,
    /// which neither Obsidian nor Quartz can follow. The rename would have
    /// broken every Markdown-style link into the folder, in a teacher's own
    /// pages, silently. Found by adversarial review 2026-09-06 and reproduced
    /// before being fixed; the mac had the identical rule and the identical
    /// defect, and took this fix unchanged.</para>
    ///
    /// <para>So: in a Markdown link, encode when the NEW name needs it,
    /// whatever the old one looked like. In a wikilink, keep the old style —
    /// <c>[[All Tasks/Quiz 1]]</c> is exactly how Obsidian writes a wikilink
    /// with a space in it, and escaping there would be the mirror-image
    /// mistake.</para>
    ///
    /// <para><b>Both branches encode the same way, and that is not an
    /// accident.</b> The second one — a segment that ARRIVED percent-encoded
    /// goes back percent-encoded, so a link a teacher already had keeps the
    /// shape it had — is the older of the two rules and was kept rather than
    /// replaced. It used <c>Uri.EscapeDataString</c> too, which is why fixing
    /// only the first line would have left a name mis-spelled whenever the old
    /// segment happened to arrive encoded. The contract has a case for each
    /// branch.</para>
    /// </summary>
    private static string Spelled(string newName, bool wasEncoded, bool markdown)
    {
        if (markdown && WouldBreakAMarkdownTarget(newName)) return PercentEncoded(newName);
        return wasEncoded ? PercentEncoded(newName) : newName;
    }

    /// <summary>
    /// The name percent-encoded so that Quartz gets the folder's REAL name back
    /// out of it. See <see cref="CharactersThatSurviveQuartzUndecoded"/> for why
    /// the allowed set is neither <c>Uri.EscapeDataString</c>'s nor the whole
    /// reserved set — both were tried on the mac and both are wrong here.
    ///
    /// <para>The walk is over UTF-8 BYTES rather than characters, deliberately.
    /// A character loop would have to encode each <c>char</c> on its own, which
    /// splits a surrogate pair and produces two invalid sequences for one
    /// letter. Every byte at or above <c>0x80</c> is outside the allowed set by
    /// construction — the set is all ASCII — so testing the byte is exact.</para>
    ///
    /// <para>Upper-case hex, matching what the contract's cases expect and what
    /// the mac's <c>addingPercentEncoding</c> writes.</para>
    /// </summary>
    private static string PercentEncoded(string name)
    {
        var built = new StringBuilder(name.Length);
        foreach (byte value in Encoding.UTF8.GetBytes(name))
        {
            char character = (char)value;
            if (value < 0x80 && CharactersThatSurviveQuartzUndecoded.IndexOf(character) >= 0)
                built.Append(character);
            else
                built.Append('%').Append(value.ToString("X2", CultureInfo.InvariantCulture));
        }
        return built.ToString();
    }

    /// <summary>
    /// Whether this name, dropped into a Markdown link destination unescaped,
    /// would end the destination early or misparse it.
    ///
    /// <para>Whitespace and round brackets, and nothing else. <b>A lone
    /// <c>%</c> was on this list for an afternoon and was taken back off</b>,
    /// so do not put it back. The argument for it was that Quartz resolves an
    /// internal link with JavaScript's <c>decodeURI</c> and
    /// <c>decodeURI("10%/Quiz.md")</c> throws — true of <c>decodeURI</c> on its
    /// own, and irrelevant here, because Quartz never sees a bare <c>%</c>: the
    /// Markdown parser normalises it to <c>%25</c> on the way to HTML long
    /// before the link transformer runs. Measured in the running image
    /// 2026-09-06, <c>[b](Top10%/Quiz.md)</c> arrives as
    /// <c>Top10%25/Quiz.md</c>. A <c>%</c> is still ENCODED whenever escaping
    /// runs for another reason, which is what stops "Top 100%" becoming the
    /// half-escaped <c>Top%20100%</c>.</para>
    /// </summary>
    private static bool WouldBreakAMarkdownTarget(string name)
    {
        foreach (char character in name)
            if (char.IsWhiteSpace(character) || character == '(' || character == ')') return true;
        return false;
    }

    private static bool NamesTheFolder(string target, string folderName)
    {
        if (PointsOutsideTheCourse(target)) return false;
        var segments = target.Split('/');
        for (int i = 0; i < segments.Length - 1; i++)
            if (SegmentIs(segments[i], folderName)) return true;
        return false;
    }

    /// <summary>
    /// Whether this path segment IS the folder — the whole segment, never a
    /// part of it, compared with percent-encoding undone and case ignored.
    /// </summary>
    private static bool SegmentIs(string segment, string folderName)
    {
        if (segment.Length == 0) return false;
        if (segment == "." || segment == "..") return false;
        return Decoded(segment).Equals(folderName, StringComparison.OrdinalIgnoreCase);
    }

    private static bool WasEncoded(string segment) => !Decoded(segment).Equals(segment, StringComparison.Ordinal);

    private static string Decoded(string segment)
    {
        if (!segment.Contains('%')) return segment;
        try { return Uri.UnescapeDataString(segment); }
        catch (UriFormatException) { return segment; }
    }

    /// <summary>
    /// A target this rename has no business touching: anything with a URL
    /// scheme, and anything absolute.
    ///
    /// <para>A relative path with a leading <c>./</c> is NOT outside the course
    /// and is still rewritten.</para>
    /// </summary>
    private static bool PointsOutsideTheCourse(string target)
    {
        if (target.Length == 0) return true;
        if (target.StartsWith("/", StringComparison.Ordinal)) return true;
        if (target.StartsWith(@"\", StringComparison.Ordinal)) return true;
        // A Windows drive letter reads as a one-character scheme, so it is
        // already caught below — but say so explicitly rather than relying on
        // that coincidence.
        if (target.Length >= 2 && char.IsLetter(target[0]) && target[1] == ':') return true;
        return Scheme.IsMatch(target);
    }
}
